c10/core/TensorImpl.cpp

```diff
@@ -560,56 +560,82 @@ SymBool TensorImpl::compute_non_overlapping_and_dense(identity<SymBool>) const {
 }
 
 // Glue compute
-// NB: intentionally not using bitwise operators.  Using bitwise operators
-// currently impedes ShapeEnv from getting crucial equalities which cause
+// NB: this logic very intentionally short circuits if possible.  Without
+// short circuiting, it causes
 // python test/functorch/test_aotdispatch.py -k
 // test_aot_autograd_symbolic_exhaustive_nn_functional_unfold_cpu_float32 to run
-// very slowly.  I think probably we just need to be able to reason through
-// And/Or, and then we can switch these to be symbolic.
+// very slowly.
+
+static bool definitely_true(SymBool b) {
+  return b.has_hint() && b.guard_bool(__FILE__, __LINE__);
+}
+
+static bool definitely_false(SymBool b) {
+  return b.has_hint() && !b.guard_bool(__FILE__, __LINE__);
+}
 
 SymBool TensorImpl::compute_is_non_overlapping_and_dense_dim4(
     identity<SymBool> type_id) {
-  return extra_meta_->is_contiguous_.guard_bool(__FILE__, __LINE__) ||
-      extra_meta_->is_channels_last_contiguous_.guard_bool(
-          __FILE__, __LINE__) ||
-      compute_non_overlapping_and_dense(type_id).guard_bool(__FILE__, __LINE__);
+  if (definitely_true(extra_meta_->is_contiguous_)) {
+    return true;
+  }
+  if (definitely_true(extra_meta_->is_channels_last_contiguous_)) {
+    return true;
+  }
+  return extra_meta_->is_contiguous_ |
+      extra_meta_->is_channels_last_contiguous_ |
```

Comment:
hmm is there a reason for the change from logical or to bitwise or? (Do we not want to use || on SymBools?)

Reply:
Oh, from looking through the next PR, it looks like our python sympy impl eventually calls `eval_is_non_overlapping_and_dense()`, which treats the output as an int.

Reply:
Logical or is not overrideable, bitwise or goes into `sym_or` which is branchless.